content/blog/2018/02/2018-02-22-cheetah.adoc

```diff
@@ -0,0 +1,148 @@
+:tags:
+- pipeline
+- performance
+- scalability
+:author: svanoort
+---
+
+:toc: 
+
+[NOTE]
+====
+This is a guest post by link:https://github.com/svanoort[Sam Van Oort],
+Software Engineer at link:http://cloudbees.com[CloudBees], and contributor to
+the Jenkins project, and maintainer of the Pipeline Plugins.
+====
+
+Since it launched, Pipeline has had a bit of a Dr. Jekyll and Mr. Hyde performance problem.  In certain circumstances, Pipeline can turn from a mild-mannered CI/CD assistant into a monster.  It will happily eat storage read/write capacity like popcorn without caring about the other concerns of our friendly butler.  When combined with other additional factors, this can result in real-world stability problems.  For example, combining slow storage with a spike in running Pipelines has brought down production Jenkins at more than one organization.  Similarly, users see issues if a busy master gets hit with an extra source of stress; past culprits have been heavy (ab)use of automated APIs, now-solved user lookup bugs, backup jobs, and plugins run amuck that load excessive numbers of builds.  Symptoms ranged from visible slowdowns in the UI to unresponsive jobs and "hung" masters.
+
+Now I'm _not_ saying this to scare people or to criticize the technology we've built. Implementing link:/blog/2017/02/01/pipeline-scalability-best-practice/[Pipeline scalability best practices] coupled with SSD storage keeps Jenkins in a happy place.  We just need context on the weaknesses to see why it's important to address them.
+
+== Introducing "Project Cheetah"
+
+Today we're announcing the first major results of "Project Cheetah", our long-running effort to address these challenges and improve Pipeline scalability.  More broadly, Cheetah aims to help in 3 places:
+
+* *Small-scale containers:* Pipeline needs to run leanly in resource-constrained containers, to enable easy scale-out without consuming excessive resources on shared container hosts.
+* *Enterprise systems:* Pipeline needs to effectively serve high-scale Jenkins instances that are central to many large companies.
+* *General case:* run Pipelines a bit more quickly on average, and allow users to get much-stronger performance in worst-case scenarios.
+
+These changes are implemented across many of the Pipeline plugins.
+
+== Yes, but what does it DO? 
+
+Project Cheetah offers several things, but the most important is Durability Settings for all Pipelines, and especially the Performance-Optimized setting.  This setting avoids several potentially unexpected performance "surprises" that may strike users.  In the general case, it greatly reduces the disk IO needs for Pipeline.  How much?  Below is a graph of storage utilization with legacy Pipeline versions (think early 2017) and with the latest version using the Performance-Optimized mode.  These are tested on an AWS instance backed by an EBS volume provisioned with 300 IOPs. 
```

Comment:
Is it possible to link to or describe the Pipeline being tested here?

Reply:
@orrc I'd like to, but there's actually a lot more complexity underlying these "simple" graphs than it would seem -- running through the methodology properly would double the length of the post, which is already pretty meaty.  Kind of had hoped to make a separate blog post of it or bring it up at Jenkins World. 

The TL;DR is that the benchmark here is a composite of 7 different Pipelines representing a mix of synthetic and realistic scenarios.  For each testcase, we gradually ramp-up the number of running Pipelines of that type until we hit a specified maximum, hold that for a few minutes (so we can collect metrics at peak), then let any running Pipelines die off.  We allow a little time for runs to complete, and then kill any that are still active... and then it's on to the next case.

Everything was Dockerized and running self-contained in AWS, with no external dependencies and the containers distributed logically across instances so that they don't interfere.  I controlled for metrics-collection overheads (minimal), and monitored that the Git Repo and Build Agents had enough resources that they wouldn't be a bottleneck.  

Basically I took pains to ensure that the only variable here is the performance of the Jenkins Master itself.  And yeah, that took a while to get all set up.